javatests/arcs/android/systemhealth/testapp/StorageCore.kt

```diff
@@ -278,73 +280,93 @@ class StorageCore(val context: Context, val lifecycle: Lifecycle) {
-                        "singletonHandle$taskId",
-                        HandleMode.ReadWrite,
-                        HandleContainerType.Singleton,
-                        TestEntity.Companion
-                    ),
-                    when (settings.storageMode) {
-                        TestEntity.StorageMode.PERSISTENT -> TestEntity.singletonPersistentStorageKey
-                        else -> TestEntity.singletonInMemoryStorageKey
+            val handle = taskHandle.handleManager.createHandle(
+                HandleSpec(
+                    "singletonHandle$taskId",
+                    HandleMode.ReadWrite,
+                    HandleContainerType.Singleton,
+                    TestEntity.Companion
+                ),
+                when (settings.storageMode) {
+                    TestEntity.StorageMode.PERSISTENT -> TestEntity.singletonPersistentStorageKey
+                    else -> TestEntity.singletonInMemoryStorageKey
+                }
+            ) as ReadWriteSingletonHandle<TestEntity>
+
+            taskHandle.handle = withContext(handle.dispatcher) {
+                val elapsedTime = measureTimeMillis { handle.awaitReady() }
+                if (settings.function == Function.LATENCY_BACKPRESSURE_TEST) {
+                    taskManagerEvents.writer.withLock {
+                        taskManagerEvents.queue.add(
+                            TaskEvent(TaskEventId.HANDLE_AWAIT_READY_TIME, elapsedTime))
                     }
-                ).awaitReady() as? ReadWriteSingletonHandle<TestEntity>)?.apply {
-                    onUpdate {
-                        entity ->
-                        if (settings.function == Function.LATENCY_BACKPRESSURE_TEST &&
-                            taskType == TaskType.WRITER
-                        ) {
-                            tasksEvents[taskId]?.writer?.withLock {
-                                tasksEvents[taskId]?.queue?.add(
-                                    TaskEvent(
-                                        TaskEventId.HANDLE_STORE_END,
-                                        System.currentTimeMillis(),
-                                        entity?.number
-                                    )
+                }
+
+                handle.onUpdate {
+                    entity ->
+                    if (settings.function == Function.LATENCY_BACKPRESSURE_TEST &&
+                        taskType == TaskType.WRITER
+                    ) {
+                        tasksEvents[taskId]?.writer?.withLock {
+                            tasksEvents[taskId]?.queue?.add(
+                                TaskEvent(
+                                    TaskEventId.HANDLE_STORE_END,
+                                    System.currentTimeMillis(),
+                                    entity?.number
                                 )
-                            }
+                            )
                         }
                     }
                 }
```

Comment:
This part doesn't need to occur on the dispatcher, callback registration is thread-safe.